### ISSUE TITLE
Memory problems when calling Taro.readxl a second time 

### ISSUE THREAD

**Issue by mbeltagy** [aviks/Taro.jl#17]
I ran into this issue when I tried to using Taro for reading a pretty large sheet. It reads well the first time around. On the second reading of the same excel file, but a different sheet I get the error below. If I start a fresh julia session and switching the reading order, it works fine on the first read and crashed on the second. The sheets I am reading are in order of 12500x10, with the first two columns being large strings, and the rest integers. 

My version information is 
```
Julia Version 0.3.7
Commit cb9bcae* (2015-03-23 21:36 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i5-2410M CPU @ 2.30GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3
```

```julia
Exception in thread "main" java.lang.OutOfMemoryError: GC overhead limit exceeded
        at java.lang.String.substring(String.java:1913)
        at org.apache.poi.ss.util.CellReference.separateRefParts(CellReference.java:364)
        at org.apache.poi.ss.util.CellReference.<init>(CellReference.java:92)
        at org.apache.poi.xssf.usermodel.XSSFCell.<init>(XSSFCell.java:106)
        at org.apache.poi.xssf.usermodel.XSSFRow.<init>(XSSFRow.java:70)
        at org.apache.poi.xssf.usermodel.XSSFSheet.initRows(XSSFSheet.java:179)
        at org.apache.poi.xssf.usermodel.XSSFSheet.read(XSSFSheet.java:143)
        at org.apache.poi.xssf.usermodel.XSSFSheet.onDocumentRead(XSSFSheet.java:130)
        at org.apache.poi.xssf.usermodel.XSSFWorkbook.onDocumentRead(XSSFWorkbook.java:286)
        at org.apache.poi.POIXMLDocument.load(POIXMLDocument.java:159)
        at org.apache.poi.xssf.usermodel.XSSFWorkbook.<init>(XSSFWorkbook.java:186)
        at org.apache.poi.ss.usermodel.WorkbookFactory.create(WorkbookFactory.java:91)
ERROR: Error calling Java: java.lang.OutOfMemoryError: GC overhead limit exceeded
 in error at error.jl:21
 in geterror at /home/user/.julia/JavaCall/src/JavaCall.jl:221
 in _jcall at /home/user/.julia/JavaCall/src/JavaCall.jl:285
 in jcall at /home/user/.julia/JavaCall/src/JavaCall.jl:235
 in readxl at /home/user/.julia/Taro/src/Taro.jl:128
 in readxl at /home/user/.julia/Taro/src/Taro.jl:103

```

**Comment by aviks** [comment]
Taro currently runs with a default JVM heap of 256M. Which seems insufficient for a file that large. I'll see if there are any leaks, but for the moment, try adding the following incantation and see if it helps

```julia
using JavaCall
using Taro
JavaCall.addOpts("-Xmx512M")
Taro.init()
```

**Comment by mbeltagy** [comment]
It tired that suggestion, and now I get a different error. 
```
signal (11): Segmentation fault
unknown function (ip: 1219780122)
unknown function (ip: 1219607234)
unknown function (ip: 1219633140)
_jcall at /home/user/.julia/JavaCall/src/JavaCall.jl:284
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
jl_f_apply at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
jcall at /home/user/.julia/JavaCall/src/JavaCall.jl:245
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
readxl at /home/user/.julia/Taro/src/Taro.jl:154
readxl at /home/user/.julia/Taro/src/Taro.jl:103
julia_readxl_20781 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: 1453079432)
unknown function (ip: 1453075600)
unknown function (ip: 1453076326)
unknown function (ip: 1453080970)
unknown function (ip: 1453074280)
unknown function (ip: 1453144135)
jl_f_top_eval at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
eval_user_input at REPL.jl:53
jlcall_eval_user_input_20175 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
anonymous at task.jl:95
jl_handle_stack_switch at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
julia_trampoline at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: 4199613)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 4199667)
unknown function (ip: 0)
Segmentation fault (core dumped)
```

I am using Oracle JDK 8,

**Comment by JonWel** [comment]
Hello,
I didn't have the out of memory error but I have the segfault.
I'm running Julia 0.3 on Ubuntu 64bit (both up to date with the ppa repos).

**Comment by aviks** [comment]
@JonWel what JVM version are you using? Is it JDK8? 

Can either of you try with JDK7 and see if you get the same segfault? Alternatively, is it possible for either of you to share the file you are using?

**Comment by JonWel** [comment]
I think I'm already using the JDK 7:
openjdk-7-jre 7u79-2.5.5-0ubuntu1

```bash
johndoe@Computer:~$ java -version
java version "1.7.0_79"
OpenJDK Runtime Environment (IcedTea 2.5.5) (7u79-2.5.5-0ubuntu1)
OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)
```

The files are generated from:
http://www.elia.be/en/grid-data/power-generation/Solar-power-generation-data/Graph
With data for a month (`BeginDate:01-01-2014` `EndDate:31-01-2014` then `refresh` and finally `export`)

I use the following code to load them:
```Julia
Taro.readxl(filePath, "A4:F2980")
```

**Comment by mbeltagy** [comment]
I switched to JDK 7: 
```
$ java -version
java version "1.7.0_79"
OpenJDK Runtime Environment (IcedTea 2.5.5) (7u79-2.5.5-0ubuntu0.14.04.2)
OpenJDK 64-Bit Server VM (build 24.79-b02, mixed mode)
```
Now it crashes on the 3rd read and not the second. 
```
signal (11): Segmentation fault
unknown function (ip: 340881130)
unknown function (ip: 340708242)
unknown function (ip: 340734148)
_jcall at /home/elbeltagy/.julia/JavaCall/src/JavaCall.jl:284
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
jl_f_apply at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
jcall at /home/elbeltagy/.julia/JavaCall/src/JavaCall.jl:245
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
readxl at /home/elbeltagy/.julia/Taro/src/Taro.jl:156
readxl at /home/elbeltagy/.julia/Taro/src/Taro.jl:103
julia_readxl_20814 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: 574180232)
unknown function (ip: 574176400)
unknown function (ip: 574177126)
unknown function (ip: 574181770)
unknown function (ip: 574175080)
unknown function (ip: 574244935)
jl_f_top_eval at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
eval_user_input at REPL.jl:53
jlcall_eval_user_input_20176 at  (unknown line)
jl_apply_generic at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
anonymous at task.jl:95
jl_handle_stack_switch at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
julia_trampoline at /usr/bin/../lib/x86_64-linux-gnu/julia/libjulia.so (unknown line)
unknown function (ip: 4199613)
__libc_start_main at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
unknown function (ip: 4199667)
unknown function (ip: 0)
Segmentation fault (core dumped)
```

**Comment by mbeltagy** [comment]
I tried @JonWel 's file and replicated his crash on the second read using JDK 7 &  JDK 8.

**Comment by aviks** [comment]
I've been working on this, and  I think I have a handle on this. However, I want to validate my hypothesis on environments other than mine.   Can I ask you to run this again with line `428` of `JavaCall.jl` commented out? The line currently reads
 `ccall(jnifunc.DeleteLocalRef, Void, (Ptr{JNIEnv}, Ptr{Void}), penv, x.ptr)` . 

Also, use 
`JavaCall.addOpts("-Xmx512M"); Javacall.addOpts("-verbose:gc:)` before `Taro.init()`

Thanks!

**Comment by JonWel** [comment]
Hello,
Commenting the line is `JavaCall.jl` do allows the other runs to succeed.
I didn't need the JavaCall options to have it right.

My file seems sufficiently small to not require `JavaCall.addOpts("-Xmx512M")`, so I didn't test with it.

I couldn't see any difference using `JavaCall.addOpts("-verbose:gc:")`. If the line `428` is not commented, it crash the same way. If the line is commented, It works the same way (no warning or anything).

**Comment by aviks** [comment]
@JonWel Thanks!

Using `vergose:gc` is only to see logs for the garbage collection. It does not change the behavior. 

So what seems to be happening is that some object that shouldn't be garbage collected is being freed. Then, any subsequent access to them causes a seg-fault. I'm debugging this right now... thanks for your help with this.

**Comment by JonWel** [comment]
I mean that I do not see any garbage collector logs in the console before the segfault (for the case with the line `428` not commented but with the `verbose:gc`). I would have expected some outputs with a `verbose` option.

**Comment by aviks** [comment]
The segmentation fault should now be fixed using *JavaCall* `v0.2.3` or higher.
